File: sdks/typescript/test/enrichment.test.ts
import { describe, expect, it } from 'vitest';

import { LogEnrichment } from '../src/enrichment.js';

describe('LogEnrichment', () => {
  it('injects job_key and runner_id into fields', () => {
    const e = new LogEnrichment('billing:invoice', 'runner-abc', []);
    const out = e.enrich({ message: 'hello' });
    expect(out.fields).toEqual({ job_key: 'billing:invoice', runner_id: 'runner-abc' });
    expect(out.message).toBe('hello');
  });

  it('preserves the original event level', () => {
    const e = new LogEnrichment('j', 'r', []);
    expect(e.enrich({ level: 'warn', message: 'hi' }).level).toBe('warn');
  });

  it('serializes tags as a JSON array string', () => {
    const e = new LogEnrichment('j', 'r', ['env=prod', 'lang=ts']);
    const out = e.enrich({ message: 'x' });
    expect(out.fields?.runner_tags).toBe('["env=prod","lang=ts"]');
  });

  it('omits runner_tags when tags is empty', () => {
    const e = new LogEnrichment('j', 'r', []);
    const out = e.enrich({ message: 'x' });
    expect(out.fields).not.toHaveProperty('runner_tags');
  });

  it('caller-provided keys win', () => {
    const e = new LogEnrichment('billing:invoice', 'runner-abc', []);
    const out = e.enrich({
      message: 'x',
      fields: { job_key: 'overridden', custom: 'value' },
    });
    expect(out.fields?.job_key).toBe('overridden');
    expect(out.fields?.runner_id).toBe('runner-abc');
    expect(out.fields?.custom).toBe('value');
  });

  it('does not mutate the source event', () => {
    const e = new LogEnrichment('j', 'r', []);
    const source = { message: 'x', fields: { a: 'b' } };
    e.enrich(source);
    expect(source.fields).toEqual({ a: 'b' });
  });
});

File: sdks/go/enrichment_test.go
package croniq

import (
	"encoding/json"
	"testing"
)

func TestEnrichEventInjectsAllThreeFieldsWhenTagsPresent(t *testing.T) {
	tags := serializeTags([]string{"env=prod", "team=ops"})
	out := enrichEvent(WorkEvent{Level: "info", Message: "hello"}, "billing:invoice", "shell-runner-1", tags)

	if got := out.Fields["job_key"]; got != "billing:invoice" {
		t.Errorf("job_key = %q, want billing:invoice", got)
	}
	if got := out.Fields["runner_id"]; got != "shell-runner-1" {
		t.Errorf("runner_id = %q, want shell-runner-1", got)
	}
	if got := out.Fields["runner_tags"]; got != `["env=prod","team=ops"]` {
		t.Errorf("runner_tags = %q, want JSON array", got)
	}
}

func TestEnrichEventSkipsRunnerTagsWhenNone(t *testing.T) {
	out := enrichEvent(WorkEvent{Level: "info", Message: "hello"}, "billing:invoice", "shell-runner-1", "")
	if _, ok := out.Fields["runner_tags"]; ok {
		t.Errorf("runner_tags should not be set when runner has no tags")
	}
	if got := out.Fields["runner_id"]; got != "shell-runner-1" {
		t.Errorf("runner_id = %q, want shell-runner-1", got)
	}
}

func TestEnrichEventPreservesCallerProvidedFields(t *testing.T) {
	in := WorkEvent{
		Level:   "warn",
		Message: "hi",
		Fields: map[string]string{
			"job_key":   "explicit:value",
			"runner_id": "explicit-runner",
		},
	}
	tags := serializeTags([]string{"env=prod"})
	out := enrichEvent(in, "auto:job", "auto-runner", tags)

	if got := out.Fields["job_key"]; got != "explicit:value" {
		t.Errorf("job_key was overwritten: got %q", got)
	}
	if got := out.Fields["runner_id"]; got != "explicit-runner" {
		t.Errorf("runner_id was overwritten: got %q", got)
	}
	if got := out.Fields["runner_tags"]; got != `["env=prod"]` {
		t.Errorf("runner_tags = %q", got)
	}

	// Original event must not be mutated.
	if _, ok := in.Fields["runner_tags"]; ok {
		t.Errorf("input event fields were mutated")
	}
}

func TestSerializeTagsEmptyReturnsBlank(t *testing.T) {
	if got := serializeTags(nil); got != "" {
		t.Errorf("got %q, want empty", got)
	}
	if got := serializeTags([]string{}); got != "" {
		t.Errorf("got %q, want empty", got)
	}
}

func TestSerializeTagsRoundTrips(t *testing.T) {
	s := serializeTags([]string{"a=1", "b=2"})
	var back []string
	if err := json.Unmarshal([]byte(s), &back); err != nil {
		t.Fatalf("unmarshal failed: %v", err)
	}
	if len(back) != 2 || back[0] != "a=1" || back[1] != "b=2" {
		t.Errorf("round-trip mismatch: %v", back)
	}
}
